### PR TITLE
feat: integrate stego-js@2

### PR DESCRIPTION
**Pull request by septs**

### CODE REVIEW

**Review by Jack-Works — APPROVED** [review]

**Comment by Jack-Works** [comment]
is this PR ready?

File: packages/maskbook/src/extension/background-script/SteganographyService/api.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    )
+    return toBuffer(data, height, width)
+}
+
+export async function decode(image: ArrayBuffer, mask: ArrayBuffer, options: DecodeOptions) {
+    return StegoAPI.decode(await toImageData(image), await toImageData(mask), options)
+}
+
+function cutImage(data: ImageData) {
+    // prettier-ignore
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/background-script/SteganographyService/api.ts:36]: ?

**Review by Jack-Works — DISMISSED** [review]

File: packages/maskbook/src/extension/background-script/SteganographyService/api.ts (hunk 1 above)
Comment (septs) [reply packages/maskbook/src/extension/background-script/SteganographyService/api.ts:36]: prettier format the code to oneline.

**Comment by Jack-Works** [comment]
@septs you need a workaround in webpack 5 see comment in #2156

**Comment by septs** [comment]
> @septs you need a workaround in webpack 5 see comment in #2156

done

File: packages/maskbook/src/extension/background-script/SteganographyService/worker.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { AsyncCall, AsyncCallOptions } from 'async-call-rpc/full'
+import { WorkerChannel } from 'async-call-rpc/utils/web/worker'
+import { serialization } from '../../../utils/type-transform/Serialization'
+import * as implementation from './worker-implementation'
+
+const options: AsyncCallOptions = {
+    channel: new WorkerChannel(),
+    strict: { methodNotFound: false },
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/background-script/SteganographyService/worker.ts:8]: Why use this flag?

File: packages/maskbook/src/extension/background-script/SteganographyService/api.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import type { AsyncCallOptions } from 'async-call-rpc'
+import { AsyncCall } from 'async-call-rpc/full'
+import { WorkerChannel } from 'async-call-rpc/utils/web/worker'
+import { serialization } from '../../../utils/type-transform/Serialization'
+import { OnDemandWorker } from '../../../web-workers/OnDemandWorker'
+
+export let StegoWorker: OnDemandWorker | undefined
+
+if (process.env.architecture) {
+    StegoWorker = new OnDemandWorker(new URL('./worker.ts', import.meta.url), { name: 'StegoWorker' })
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/background-script/SteganographyService/api.ts:12]: Please add `__webpack_public_path__` workaround I mentioned in #2156

File: packages/maskbook/src/extension/background-script/SteganographyService/api.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                resolve(loadBuffer(blob))
+            } else {
+                reject(new Error('fail to generate array buffer'))
+            }
+        }
+        canvas.toBlob(callback, 'image/png')
+    })
+}
+
+function loadBuffer(blob: Blob) {
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/background-script/SteganographyService/api.ts:77]: We have polyfill for `Blob.prototype.arrayBuffer`

**Review by Jack-Works — DISMISSED** [review]

File: packages/maskbook/src/extension/background-script/SteganographyService/api.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import type { AsyncCallOptions } from 'async-call-rpc'
+import { AsyncCall } from 'async-call-rpc/full'
+import { WorkerChannel } from 'async-call-rpc/utils/web/worker'
+import { serialization } from '../../../utils/type-transform/Serialization'
+import { OnDemandWorker } from '../../../web-workers/OnDemandWorker'
+
+export let StegoWorker: OnDemandWorker | undefined
+
+if (process.env.architecture) {
+    __webpack_public_path__ = browser.runtime.getURL('/')
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/background-script/SteganographyService/api.ts:12]: Now this line should be removed per #2179 changes

Comment (Jack-Works) [reply packages/maskbook/src/extension/background-script/SteganographyService/api.ts:12]: besides this one, everything else LGTM you can remove this line and merge it if all functions work well cc @septs

File: package.json [hunk 6]
```diff
@@ -48,7 +48,7 @@
     "@dimensiondev/kit": "0.0.0-20210115033306-5d4304b",
     "@dimensiondev/matrix-js-sdk-type": "=1.0.0-20200731085604-dc2de92",
     "@dimensiondev/metamask-extension-provider": "^3.0.0-20210125122006-86e9bae",
-    "@dimensiondev/stego-js": "=0.11.1-20201027083223-8ab41be",
+    "@dimensiondev/stego-js": "=0.13.0-20201224101931-64ee02e",
```
Comment (Jack-Works) [inline package.json:51]: Hi plz rebase to master and add this deps in packages/maskbook/package.json thanks

**Comment by Jack-Works** [comment]
seems have a serious problem in the algorithm. burning my CPU

cc @yisiliu 

image

**Comment by yisiliu** [comment]
> seems have a serious problem in the algorithm. burning my CPU
> 
> cc @yisiliu
> 
> image

Let me take a look

**Comment by Jack-Works** [comment]
FYI: `algr=V2`, `width=1200`, `height=681`, `data` is a Uint8ClampedArray of size `3268800`, `size=8`

**Comment by Jack-Works** [comment]
If you find the Worker is killed too quickly, please replace `terminate` method in `OnDemandWorker` class with empty implementation

**Review by Jack-Works — APPROVED** [review]
@septs feel free to merge (using squash) it if u think it is working correctly. code LGTM

**Comment by Tedko** [comment]
@guanbinrui take a look what do u think

**Pull request closed without merging**